File: Backend/src/modules/orders/models/order.model.ts
import { Schema, model, Types, type Document, type Model } from 'mongoose';

export enum OrderStatus {
  Pending = 'pending',
  InProgress = 'in_progress',
  Completed = 'completed',
  Cancelled = 'cancelled',
}

export interface IOrder {
  orderNumber: string;
  customerId: Types.ObjectId;
  quantity: number;
  totalPrice: number;
  deadline: Date;
  status: OrderStatus;
}

export interface IOrderDocument extends IOrder, Document {
  createdAt: Date;
  updatedAt: Date;
}

const orderSchema = new Schema<IOrderDocument>(
  {
    orderNumber: {
      type: String,
      required: true,
      unique: true,
      trim: true,
      uppercase: true,
      maxlength: 40,
      index: true,
    },
    customerId: {
      type: Schema.Types.ObjectId,
      ref: 'Customer',
      required: true,
      index: true,
    },
    quantity: {
      type: Number,
      required: true,
      min: 1,
    },
    totalPrice: {
      type: Number,
      required: true,
      min: 0,
    },
    deadline: {
      type: Date,
      required: true,
      index: true,
    },
    status: {
      type: String,
      enum: Object.values(OrderStatus),
      default: OrderStatus.Pending,
      required: true,
      index: true,
    },
  },
  {
    timestamps: true,
    versionKey: false,
  },
);

orderSchema.index({ customerId: 1, status: 1 });
orderSchema.index({ status: 1, deadline: 1 });

export const OrderModel: Model<IOrderDocument> = model<IOrderDocument>('Order', orderSchema);

File: lib/features/orders/data/models/order_model.dart
import '../../../workshop/models/workshop_mock_models.dart';

final class OrderModel {
  const OrderModel({
    required this.id,
    required this.orderNumber,
    required this.customerId,
    required this.quantity,
    required this.totalPrice,
    required this.deadline,
    required this.status,
    required this.createdAt,
    required this.updatedAt,
  });

  final String id;
  final String orderNumber;
  final String customerId;
  final int quantity;
  final double totalPrice;
  final DateTime deadline;
  final String status;
  final DateTime createdAt;
  final DateTime updatedAt;

  factory OrderModel.fromJson(Map<String, dynamic> json) {
    return OrderModel(
      id: json['id'] as String,
      orderNumber: json['orderNumber'] as String,
      customerId: json['customerId'] as String,
      quantity: json['quantity'] as int,
      totalPrice: (json['totalPrice'] as num).toDouble(),
      deadline: DateTime.parse(json['deadline'] as String),
      status: json['status'] as String,
      createdAt: DateTime.parse(json['createdAt'] as String),
      updatedAt: DateTime.parse(json['updatedAt'] as String),
    );
  }

  WorkshopOrder toWorkshopOrder() {
    return WorkshopOrder(
      id: id,
      productName: orderNumber,
      orderedQty: quantity,
      producedQty: status == 'completed' ? quantity : 0,
      deadline: deadline,
      note: 'Customer: $customerId',
    );
  }
}
